File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = [0]
        def countNodes(root, val, count):
            if root == None:
                return
            if root.val >= val:
                count[0] += 1
            countNodes(root.left, root.val if root.val > val else val, count)
            countNodes(root.right, root.val if root.val > val else val, count)

        countNodes(root, root.val, count)
        return count[0]

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null) return false;
        if(root.val == subRoot.val && isSame(root, subRoot)) {
            return true;
        }
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    private boolean isSame(TreeNode p1, TreeNode p2) {
        if(p1 == null && p2 == null) {
            return true;
        }
        else if(p1 != null && p2 != null) {
            if(p1.val == p2.val) {
                return isSame(p1.left, p2.left) && isSame(p1.right, p2.right);
            }
        }
        return false;
    }
}
